### DISCUSSION TITLE
isosurface alpha

### DISCUSSION THREAD

**Discussion by AshtonSBradley** [JuliaPlots/Makie.jl#4202] [Q&A]
There is an unexpected behavior with respect to alpha of topologically interesting surfaces:

```
using GLMakie
using ColorSchemes
cmap = get(colorschemes[:plasma], LinRange(0,1,100))
	cmap1 = [(cmap[i],0.65) for i in 1:100]
	ξ = -40:1:40
	c = 20
	a = 15
	torus = [((hypot(X,Y)-c)^2+Z^2-a^2) for X in ξ, Y in ξ, Z in ξ]
	volume(torus, algorithm = :iso,isovalue =100,isorange =50, colormap=cmap1)
	current_figure()
```

gives 

<img width="470" alt="Screen Shot 2021-08-21 at 11 08 57 PM" src="https://user-images.githubusercontent.com/24194258/130319910-aaafe1e6-c1b3-4128-b5c7-36d8bb1b677e.png">

Shouldn't the alpha allow the interior of the torus to be partially visible? At present it means that interesting topology can't be visualised. 

For simple distinct surfaces, the result is more intuitive, in that the surfaces are visible through each other:

```
using GLMakie, ColorSchemes
cbarPal= :plasma
cmap = get(colorschemes[cbarPal], LinRange(0,1,100))
cmap2 = [(cmap[i],0.65) for i in 1:100]

X = -40:5:40
ellipse1 = [X^2/2 + Y^2 + 2Z^2 for X in X, Y in X, Z in X];
ellipse2 = [X^2/2 + (Y-20)^2 + 2Z^2 for X in X, Y in X, Z in X];
volume(ellipse2, algorithm = :iso,isovalue =350,isorange = 50, colormap=cmap2)
volume!(ellipse1, algorithm = :iso, isovalue =700,isorange = 50, colormap=cmap2)
current_figure()
```

<img width="435" alt="Screen Shot 2021-08-21 at 11 11 50 PM" src="https://user-images.githubusercontent.com/24194258/130319970-ed561472-11dc-4b98-93da-c98e2463f5ff.png">

This was discussed over at [discourse ](https://discourse.julialang.org/t/makie-isosurface-transparency/65214/9)

**Comment by SimonDanisch** [comment]
Try `contour`, that gives you more knobs to play around with:
```julia
using ColorSchemes
cmap = get(colorschemes[:plasma], LinRange(0,1,100))
cmap1 = [RGBAf0(cmap[i],0.65) for i in 1:100]
ξ = -40:1:40
c = 20
a = 15
torus = [((hypot(X,Y)-c)^2+Z^2-a^2) for X in ξ, Y in ξ, Z in ξ]
contour(torus, levels=[100.0, 500.0], alpha=0.05)
current_figure() |> display
```
![image](https://user-images.githubusercontent.com/1010467/130320424-7eaea6db-b930-4081-9b37-6412ad3d50a3.png)

**Comment by AshtonSBradley** [comment]
It's quite murky compared to iso. Is this really a way forward here? I would have thought alpha should apply to the whole scene including between layers in the current surface. Is there a fundamental rendering approach that isn't going to be changed any time soon?

**Comment by SimonDanisch** [comment]
>Is there a fundamental rendering approach that isn't going to be changed any time soon?

What do you mean?

>It's quite murky compared to iso

Iso doesn't support alpha right now, since it stops at the isosurface.  What do you mean by murky? Have you tried tweaking the parameters? `contour(vol)` is pretty much the iso algorithm with alpha support...

**Comment by AshtonSBradley** [comment]
I don't know how to get the result I want by tweaking. This is as close as I can get. Still feels quite murky, probably just my ignorance about how to tweak. Surface contours visible and I don't know how to turn on an external light source to get some surface shine (which seems to be default in iso?). Also, doesn't seem to respond to changing colormaps:

```julia
cmap3 = get(colorschemes[:plasma], LinRange(0,1,100))
ξ = -40:1:40
c = 20
a = 15
torus = [((hypot(X,Y)-c)^2+Z^2-a^2) for X in ξ, Y in ξ, Z in ξ]
contour(torus, alpha=0.13,levels=[300],colormap=cmap3,colorrange=[0,800])
current_figure()
```
<img width="590" alt="Screen Shot 2021-08-27 at 8 47 56 AM" src="https://user-images.githubusercontent.com/24194258/131034228-941e8c6d-3e80-4f94-872e-90ecbe0d4bce.png">

whereas I am aiming for something like this where the top is clearly the top, and the underside is visibly obscured by the other surfaces, but still visible. 

<img width="569" alt="Screen Shot 2021-08-27 at 8 51 58 AM" src="https://user-images.githubusercontent.com/24194258/131034437-e09c6b72-4ed7-4a73-b151-8dc4127b5212.png">

**Comment by SimonDanisch** [comment]
Hm, seems like the normals are inverted for contours...When I flip them I get:
![image](https://user-images.githubusercontent.com/1010467/131037152-6380e563-665b-41b9-8303-602be062918a.png)

Btw, the image you post looks like a transparent mesh, and not the kind of ray tracing that the volume algorithms do..
You may want to do something like:
```julia
using Meshing, GeometryBasics
algo = MarchingCubes(iso=100, insidepositive=true)
mc = GeometryBasics.Mesh(torus, algo)
Makie.mesh(normal_mesh(mc), 
    color=(:red, 0.5), 
    # Light comes from (0, 0, 15), i.e the sphere
    lightposition = Vec3f(20, 0,-20),
    # base light of the plot only illuminates red colors
    ambient = Vec3f(0.1),
    # light from source (sphere) illuminates yellow colors
    diffuse = Vec3f(0.5),
    # reflections illuminate blue colors
    specular = Vec3f(0.8),
    # Reflections are sharp
    shininess = 50f0)|> display
 ```
![image](https://user-images.githubusercontent.com/1010467/131038687-adc5ff41-c5ca-4338-b510-537d9ae0eb01.png)

**Comment by AshtonSBradley** [comment]
Thanks a lot. I've got it to here, which I'm pretty happy with. 

Given the fairly non-standard usage, is there somewhere this would like to live in docs? 

Or an advanced example @lazarusA ?

```julia	
	using GLMakie
	using Meshing, GeometryBasics
	ξ = -40:.5:40
	c = 20
	a = 15
	torus = [((hypot(X,Y)-c)^2+Z^2-a^2) for X in ξ, Y in ξ, Z in ξ]
	
	algo = MarchingCubes(iso=100, insidepositive=false)
	mc = GeometryBasics.Mesh(torus, algo)
	figt = Figure(resolution=(1200,1200))
	Makie.mesh(figt[1,1],normal_mesh(mc), 
	    color=(:red, 0.6), 
	    # Light comes from (0, 0, 15), i.e the sphere
	    lightposition = Vec3f0(20, 120,100),
	    # base light of the plot only illuminates red colors
	    ambient = Vec3f0(0.3),
	    # light from source (sphere) illuminates yellow colors
	    diffuse = Vec3f0(0.8),
	    # reflections illuminate blue colors
	    specular = Vec3f0(0.8),
	    # Reflections are sharp
	    shininess = 40f0)
	figt
```

<img width="718" alt="Screen Shot 2021-08-29 at 1 41 35 PM" src="https://user-images.githubusercontent.com/24194258/131235436-206acef0-6ce2-4b7f-b124-d7d8cbeff8b8.png">

**Comment by AshtonSBradley** [comment]
to light up the tube better (i.e. the front), I want to put more than one light source. 
Is it possible?  

If I try to add a second light source using the same syntax I get the error 
```

syntax: keyword argument "lightposition" repeated in call to "Makie.mesh" around /Users/abradley/Dropbox/Julia/MakieExamples/figure_play.plu.jl#==#c4b7b098-ee54-40dd-84bb-066ab3e6af2a:12
```

**Comment by ffreyer** [comment]
There is a `backlight` attribute that acts as a multiplier to a light coming from the opposite direction. That's all you can do at the moment.
